### PR TITLE
Clean up and optimize byte<->float and Rgba32 <-> Vector4 conversion

### PR DESCRIPTION
**Pull request by antonfirsov**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
- Uniformized `Span<byte>` -> `Span<float>` (and opposite) conversion methods in `SimdUtils` which would be useful for #733.
- Changed `Rgba32.PixelOperations` to consume these uniformized converters in bulk conversions to/from `Vector4`
- New implementations for the conversions based on dotnet/corefx#15957. These are only accelerated by the newest RyuJIT deployed with .NET Core 2.1 and .NET Framework 4.7.2. Currently the latter is not being targeted directly by ImageSharp, so the new implementations are only used on .NET Core 2.1
  - `Span<byte>` -> `Span<float>` (thus `Span<Rgba32>` -> `Span<Vector4>`)  is **3x faster** than the current implementation on `main`
  - The opposite direction is ~2x faster

### Benchmark results
Detailed benchmark results can be found in comments in benchmark code ([ToVector4_Rgba32](https://github.com/SixLabors/ImageSharp/blob/520c6fc564c7748f73e9e7e64c483f48f0e2490f/tests/ImageSharp.Benchmarks/Color/Bulk/ToVector4.cs#L192), [PackFromVector4_Rgba32](https://github.com/SixLabors/ImageSharp/blob/520c6fc564c7748f73e9e7e64c483f48f0e2490f/tests/ImageSharp.Benchmarks/Color/Bulk/PackFromVector4.cs#L101)). 

Here are the interesting bits:

1. `ToVector4_Rgba32`
```
      FallbackIntrinsics128 |    Core |  2048 | 5,017.89 ns | 4,021.533 ns | 227.2241 ns |   1.24 |     0.05 |      - |       0 B |
         BasicIntrinsics256 |    Core |  2048 | 4,046.51 ns | 1,150.390 ns |  64.9992 ns |   1.00 |     0.00 |      - |       0 B |
         ExtendedIntrinsics |    Core |  2048 | 1,130.59 ns |   832.588 ns |  47.0427 ns |!! 0.28 |     0.01 |      - |       0 B |
       PixelOperations_Base |    Core |  2048 | 6,752.68 ns |   272.820 ns |  15.4148 ns |   1.67 |     0.02 |      - |      24 B |
PixelOperations_Specialized |    Core |  2048 | 1,126.13 ns |    79.192 ns |   4.4745 ns |!! 0.28 |     0.00 |      - |       0 B |
```
2. `PackFromVector4_Rgba32`
```
      FallbackIntrinsics128 |    Core |  2048 |  6,644.65 ns | 2,677.090 ns | 151.2605 ns |   1.69 |     0.05 |      - |       0 B |
         BasicIntrinsics256 |    Core |  2048 |  3,923.70 ns | 1,971.760 ns | 111.4081 ns |   1.00 |     0.00 |      - |       0 B |
          ExtendedIntrinsic |    Core |  2048 |  2,092.32 ns |   375.657 ns |  21.2253 ns |!! 0.53 |     0.01 |      - |       0 B |
       PixelOperations_Base |    Core |  2048 | 16,875.73 ns | 1,271.957 ns |  71.8679 ns |   4.30 |     0.10 |      - |      24 B |
PixelOperations_Specialized |    Core |  2048 |  2,129.92 ns |   262.888 ns |  14.8537 ns |!! 0.54 |     0.01 |      - |       0 B |
```

### CODE REVIEW

File: src/ImageSharp/Common/Helpers/ImageMaths.cs [hunk 1]
```diff
@@ -39,14 +39,39 @@ public static int LeastCommonMultiple(int a, int b)
...
+        /// <paramref name="m"/> should be power of 2.
+        /// </summary>
+        [MethodImpl(InliningOptions.ShortMethod)]
+        public static int ModuloP2(int x, int m)
+        {
+            return x & (m - 1);
+        }
+
+        [MethodImpl(InliningOptions.ShortMethod)]
+        public static float Clamp(float x, float min, float max) => Math.Min(max, Math.Max(min, x));
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Common/Helpers/ImageMaths.cs]: You might wanna benchmark this. The `IComparableExtensions` version should be a good bit faster.

I do, however want to ditch the extension method for more clear `MathUtils.Clamp***` methods.

File: src/ImageSharp/Common/Helpers/SimdUtils.BasicIntrinsics256.cs [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+            }
+
+            /// <summary>
+            /// SIMD optimized implementation for <see cref="SimdUtils.BulkConvertByteToNormalizedFloat"/>.
+            /// Works only with `dest.Length` divisible by 8.
+            /// Implementation adapted from:
+            /// http://lolengine.net/blog/2011/3/20/understanding-fast-float-integer-conversions
+            /// http://stackoverflow.com/a/536278
+            /// </summary>
+            internal static void BulkConvertByteToNormalizedFloat(ReadOnlySpan<byte> source, Span<float> dest)
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Common/Helpers/SimdUtils.BasicIntrinsics256.cs]: Could this perhaps be private so it cannot be called without sanitation?

File: src/ImageSharp/Common/Helpers/SimdUtils.ExtendedIntrinsics.cs [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        /// See:
+        /// https://github.com/dotnet/coreclr/pull/10662
+        /// API Proposal:
+        /// https://github.com/dotnet/corefx/issues/15957
+        /// </summary>
+        public static class ExtendedIntrinsics
+        {
+            public static bool IsAvailable { get; } =
+#if NETCOREAPP2_1
+                // TODO: Also available in .NET 4.7.2, we need to add a build target!
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Common/Helpers/SimdUtils.ExtendedIntrinsics.cs]: Add one if you need one.

File: src/ImageSharp/Common/Helpers/SimdUtils.FallbackIntrinsics128.cs [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+                var half = new Vector4(0.5f);
+                var maxBytes = new Vector4(255f);
+
+                for (int i = 0; i < count; i++)
+                {
+                    Vector4 s = Unsafe.Add(ref sBase, i);
+                    s *= maxBytes;
+                    s += half;
+
+                    // I'm not sure if Vector4.Clamp() is properly implemented with intrinsics.
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Common/Helpers/SimdUtils.FallbackIntrinsics128.cs]: I do recall reading somewhere it wasn't.

File: src/ImageSharp/Common/Tuples/Octet.cs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+            [FieldOffset(6)]
+            public byte V6;
+
+            [FieldOffset(7)]
+            public byte V7;
+
+            public override string ToString()
+            {
+                return $"[{this.V0},{this.V1},{this.V2},{this.V3},{this.V4},{this.V5},{this.V6},{this.V7}]";
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Common/Tuples/Octet.cs]: I'm favoring the `TypeName(field, field, field)` format now for ToString constituency.

File: src/ImageSharp/Formats/Jpeg/Components/Decoder/ColorConverters/JpegColorConverter.FromYCbCrSimdAvx2.cs [hunk 6]
```diff
@@ -6,7 +6,7 @@
 using System.Runtime.CompilerServices;
 using System.Runtime.InteropServices;
 
-using SixLabors.ImageSharp.Common.Tuples;
+using SixLabors.ImageSharp.Tuples;
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Jpeg/Components/Decoder/ColorConverters/JpegColorConverter.FromYCbCrSimdAvx2.cs]: Would these come under our Primitives namespace?

File: src/ImageSharp/PixelFormats/PixelOperations{TPixel}.cs [hunk 7]
```diff
@@ -29,10 +29,12 @@ public partial class PixelOperations<TPixel>
         /// <param name="count">The number of pixels to convert.</param>
         internal virtual void PackFromVector4(ReadOnlySpan<Vector4> sourceVectors, Span<TPixel> destinationColors, int count)
         {
-            GuardSpans(sourceVectors, nameof(sourceVectors), destinationColors, nameof(destinationColors), count);
+            ReadOnlySpan<Vector4> sourceVectors1 = sourceVectors;
```
Comment (JimBobSquarePants) [inline src/ImageSharp/PixelFormats/PixelOperations{TPixel}.cs]: How come these are reassigned?

File: tests/ImageSharp.Benchmarks/General/BasicMath/ModuloPowerOfTwoVariable.cs [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+            return this.value % this.m;
+        }
+
+        [Benchmark]
+        public int Bitwise()
+        {
+            return ImageMaths.ModuloP2(this.value, this.m);
+        }
+
+        // RESULTS:
```
Comment (JimBobSquarePants) [inline tests/ImageSharp.Benchmarks/General/BasicMath/ModuloPowerOfTwoVariable.cs]: I need to make sure I add the results like this when benchmarking. Really good idea.

File: src/ImageSharp/Common/Helpers/ImageMaths.cs (hunk 1 above)
Comment (antonfirsov) [reply src/ImageSharp/Common/Helpers/ImageMaths.cs]: Good point, will do compare!

File: src/ImageSharp/Common/Helpers/SimdUtils.BasicIntrinsics256.cs (hunk 2 above)
Comment (antonfirsov) [reply src/ImageSharp/Common/Helpers/SimdUtils.BasicIntrinsics256.cs]: These methods are all unit tested separately (so coverage is independent from current HW configuration), so we need them as `internal`. 

I can improve the input checking a bit however.

File: src/ImageSharp/Common/Helpers/SimdUtils.ExtendedIntrinsics.cs (hunk 3 above)
Comment (antonfirsov) [reply src/ImageSharp/Common/Helpers/SimdUtils.ExtendedIntrinsics.cs]: I'm planning to do this in a separate PR, or open an up-for-grabs issue. @iamcarbon is the champion of this stuff! 😄

File: src/ImageSharp/Formats/Jpeg/Components/Decoder/ColorConverters/JpegColorConverter.FromYCbCrSimdAvx2.cs (hunk 6 above)
Comment (antonfirsov) [reply src/ImageSharp/Formats/Jpeg/Components/Decoder/ColorConverters/JpegColorConverter.FromYCbCrSimdAvx2.cs]: Definitely a better place!

File: src/ImageSharp/PixelFormats/PixelOperations{TPixel}.cs (hunk 7 above)
Comment (antonfirsov) [reply src/ImageSharp/PixelFormats/PixelOperations{TPixel}.cs]: Good catch! Result of undoing a refactor with R#. Will fix it.

File: src/ImageSharp/Common/Helpers/SimdUtils.BasicIntrinsics256.cs (hunk 2 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Common/Helpers/SimdUtils.BasicIntrinsics256.cs]: Ah right... Makes sense then. Carry on!

File: src/ImageSharp/Common/Helpers/ImageMaths.cs (hunk 1 above)
Comment (dlemstra) [reply src/ImageSharp/Common/Helpers/ImageMaths.cs]: Do we need the duplication?

File: src/ImageSharp/PixelFormats/Rgba32.PixelOperations.cs [hunk 9]
```diff
@@ -19,130 +18,31 @@ public partial struct Rgba32
...
-                }
-            }
-
             /// <inheritdoc />
             internal override void ToVector4(ReadOnlySpan<Rgba32> sourceColors, Span<Vector4> destinationVectors, int count)
             {
                 Guard.MustBeSizedAtLeast(sourceColors, count, nameof(sourceColors));
                 Guard.MustBeSizedAtLeast(destinationVectors, count, nameof(destinationVectors));
 
-                if (count < 256 || !Vector.IsHardwareAccelerated)
```
Comment (dlemstra) [inline src/ImageSharp/PixelFormats/Rgba32.PixelOperations.cs]: Do we no longer need the count check?

File: tests/ImageSharp.Benchmarks/Color/Bulk/PackFromVector4.cs [hunk 10]
```diff
@@ -40,7 +44,7 @@ public void Cleanup()
             this.source.Dispose();
         }
 
-        [Benchmark(Baseline = true)]
+        //[Benchmark]
```
Comment (dlemstra) [inline tests/ImageSharp.Benchmarks/Color/Bulk/PackFromVector4.cs]: Should this still be commented?

File: tests/ImageSharp.Benchmarks/Color/Bulk/ToVector4.cs [hunk 11]
```diff
@@ -38,33 +50,171 @@ public void Cleanup()
             this.destination.Dispose();
         }
 
-        [Benchmark(Baseline = true)]
+        //[Benchmark]
```
Comment (dlemstra) [inline tests/ImageSharp.Benchmarks/Color/Bulk/ToVector4.cs]: Should this still be commented?

File: src/ImageSharp/Common/Helpers/ImageMaths.cs (hunk 1 above)
Comment (antonfirsov) [reply src/ImageSharp/Common/Helpers/ImageMaths.cs]: The one in `ComparableExtensions` is faster, I'm removing this!

File: src/ImageSharp/PixelFormats/Rgba32.PixelOperations.cs (hunk 9 above)
Comment (antonfirsov) [reply src/ImageSharp/PixelFormats/Rgba32.PixelOperations.cs]: It was an optimization for small buffers, but the new logic made it obsolete.

File: tests/ImageSharp.Benchmarks/Color/Bulk/PackFromVector4.cs (hunk 10 above)
Comment (antonfirsov) [reply tests/ImageSharp.Benchmarks/Color/Bulk/PackFromVector4.cs]: The benchmark code serves as an information, but the execution is unnecessary, unless someone wants to evaluate that specific method in future investigations.

I wish there was some better way to Skip benchmarks without completely dropping their code.

File: src/ImageSharp/PixelFormats/Rgba32.PixelOperations.cs (hunk 9 above)
Comment (dlemstra) [reply src/ImageSharp/PixelFormats/Rgba32.PixelOperations.cs]: Okay 👍 Just wanted to make sure that you not missed this.

**Comment by antonfirsov** [comment]
@JimBobSquarePants @dlemstra all findings were addressed. Gonna merge this as soon as the compilation is finished so we can go on with #729.

**Comment by JimBobSquarePants** [comment]
Awesome!

**Pull request merged**